File: sdks/typescript/tests/job_run.test.ts
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import { describe, test } from "node:test";

import { create, type DescMessage, type MessageShape, toBinary } from "@bufbuild/protobuf";
import { timestampFromDate } from "@bufbuild/protobuf/wkt";
import { createRouterTransport, type Transport } from "@connectrpc/connect";

import { ArtifactRefSchema } from "../../../protocols/generated/typescript/artifact/v1/artifact_reference_pb.js";
import {
	AcquireAttemptLeaseRequestSchema,
	CancelAttemptRequestSchema,
	CancelJobRequestSchema,
	CommitAttemptRequestSchema,
	HeartbeatAttemptRequestSchema,
	JobService,
	ListAttemptsRequestSchema,
	ListJobsRequestSchema,
	ListRunsRequestSchema,
	RenewAttemptLeaseRequestSchema,
	RunService,
} from "../../../protocols/generated/typescript/internal/job/v1/job_service_pb.js";
import {
	AttemptSchema,
	AttemptState,
} from "../../../protocols/generated/typescript/job/v1/attempt_pb.js";
import { RequestJobCommandSchema } from "../../../protocols/generated/typescript/job/v1/job_commands_pb.js";
import { JobState } from "../../../protocols/generated/typescript/job/v1/job_pb.js";
import { LeaseFenceSchema } from "../../../protocols/generated/typescript/job/v1/lease_fencing_pb.js";
import { OperationState } from "../../../protocols/generated/typescript/operation/v1/operation_pb.js";
import { RunState } from "../../../protocols/generated/typescript/job/v1/run_pb.js";
import {
	ClientConfig,
	Environment,
	FakeRuntime,
	LeaseCredential,
	MindcladeClient,
} from "../src/index.js";

const TOKEN = `lease-token-${"s".repeat(40)}`;
const TENANT = "tenant-1";
const PROJECT = "project-1";
const PRINCIPAL = "principal-1";

const testTransport = (routes: Parameters<typeof createRouterTransport>[0]): Transport => {
	const delegate = createRouterTransport(routes);
	return {
		unary(method, signal, _timeoutMs, header, input, contextValues) {
			return delegate.unary(method, signal, undefined, header, input, contextValues);
		},
		stream(method, signal, _timeoutMs, header, input, contextValues) {
			return delegate.stream(method, signal, undefined, header, input, contextValues);
		},
	};
};

const digest = (value: string): string =>
	`sha256:${createHash("sha256").update(value).digest("hex")}`;

const canonical = <Schema extends DescMessage>(
	schema: Schema,
	value: MessageShape<Schema>,
): string => `sha256:${createHash("sha256").update(toBinary(schema, value)).digest("hex")}`;

const verifyContext = <Schema extends DescMessage>(
	schema: Schema,
	value: MessageShape<Schema> & {
		context?: {
			canonicalRequestDigest: string;
			principalId: string;
			projectId: string;
			tenantId: string;
		};
	},
): void => {
	const clone = create(schema, value) as typeof value;
	const context = clone.context;
	assert.ok(context !== undefined);
	delete clone.context;
	assert.equal(context.canonicalRequestDigest, canonical(schema, clone));
	assert.equal(context.tenantId, TENANT);
	assert.equal(context.projectId, PROJECT);
	assert.equal(context.principalId, PRINCIPAL);
};

describe("JobService and RunService ergonomic facades", () => {
	test("cover every intended RPC while keeping lease credentials metadata-only", async () => {
		const runtime = new FakeRuntime({
			now: 1_800_000_000_000,
			requestIds: Array.from({ length: 32 }, (_, index) => `request-${index}`),
		});
		const deadline = timestampFromDate(new Date(runtime.now + 60 * 60 * 1_000));
		const artifact = create(ArtifactRefSchema, {
			digest: `sha256:${"a".repeat(64)}`,
			mediaType: "application/json",
			sizeBytes: 12n,
		});
		const job = {
			jobId: "jobs/job-1",
			operationId: "operations/op-1",
			projectId: PROJECT,
			resourceVersion: 1n,
			state: JobState.RUNNING,
			tenantId: TENANT,
		};
		const operation = {
			jobId: job.jobId,
			operationId: job.operationId,
			projectId: PROJECT,
			resourceVersion: 1n,
			state: OperationState.RUNNING,
			tenantId: TENANT,
		};
		const run = {
			jobId: job.jobId,
			leaseEpoch: 1n,
			projectId: PROJECT,
			resourceVersion: 1n,
			runId: "runs/run-1",
			state: RunState.EXECUTING,
			tenantId: TENANT,
		};
		const attempt = create(AttemptSchema, {
			attemptId: "attempts/attempt-1",
			jobId: job.jobId,
			leaseEpoch: 1n,
			leaseExpiresAt: deadline,
			projectId: PROJECT,
			resourceVersion: 1n,
			runId: run.runId,
			state: AttemptState.RUNNING,
			tenantId: TENANT,
			workerId: "workers/worker-1",
		});
		const fence = create(LeaseFenceSchema, {
			attemptId: attempt.attemptId,
			deadline,
			jobId: attempt.jobId,
			leaseEpoch: attempt.leaseEpoch,
			leaseTokenDigest: digest(TOKEN),
			projectId: PROJECT,
			runId: attempt.runId,
			tenantId: TENANT,
		});
		const calls: string[] = [];
		const leaseHeaders: Array<string | null> = [];
		const transport = testTransport((router) => {
			router.service(JobService, {
				requestJob(request) {
					calls.push("RequestJob");
					assert.ok(request.command !== undefined);
					verifyContext(RequestJobCommandSchema, request.command);
					return { job, operation };
				},
				getJob(request) {
					calls.push("GetJob");
					assert.equal(request.name, job.jobId);
					return { job };
				},
				listJobs(request) {
					calls.push("ListJobs");
					assert.equal(request.parent, "tenants/tenant-1/projects/project-1");
					return { jobs: [job], page: { nextPageToken: "jobs-next" } };
				},
				cancelJob(request) {
					calls.push("CancelJob");
					verifyContext(CancelJobRequestSchema, request);
					return { operation };
				},
			});
			router.service(RunService, {
				getRun(request) {
					calls.push("GetRun");
					assert.equal(request.name, run.runId);
					return { run };
				},
				listRuns(request) {
					calls.push("ListRuns");
					assert.equal(request.parent, job.jobId);
					return { runs: [run], page: { nextPageToken: "runs-next" } };
				},
				getAttempt(request) {
					calls.push("GetAttempt");
					assert.equal(request.name, attempt.attemptId);
					return { attempt };
				},
				listAttempts(request) {
					calls.push("ListAttempts");
					assert.equal(request.parent, run.runId);
					return { attempts: [attempt], page: { nextPageToken: "attempts-next" } };
				},
				acquireAttemptLease(request, context) {
					calls.push("AcquireAttemptLease");
					verifyContext(AcquireAttemptLeaseRequestSchema, request);
					assert.equal(context.requestHeader.has("x-mindclade-lease-token"), false);
					context.responseHeader.set("x-mindclade-lease-token", TOKEN);
					return { attempt, fence };
				},
				renewAttemptLease(request, context) {
					calls.push("RenewAttemptLease");
					verifyContext(RenewAttemptLeaseRequestSchema, request);
					leaseHeaders.push(context.requestHeader.get("x-mindclade-lease-token"));
					return { attempt, fence };
				},
				heartbeatAttempt(request, context) {
					calls.push("HeartbeatAttempt");
					verifyContext(HeartbeatAttemptRequestSchema, request);
					leaseHeaders.push(context.requestHeader.get("x-mindclade-lease-token"));
					return { attempt, fence, observedAt: timestampFromDate(new Date(runtime.now)) };
				},
				cancelAttempt(request, context) {
					calls.push("CancelAttempt");
					verifyContext(CancelAttemptRequestSchema, request);
					leaseHeaders.push(context.requestHeader.get("x-mindclade-lease-token"));
					return { attempt, run };
				},
				commitAttempt(request, context) {
					calls.push("CommitAttempt");
					verifyContext(CommitAttemptRequestSchema, request);
					leaseHeaders.push(context.requestHeader.get("x-mindclade-lease-token"));
					return { attempt, run };
				},
			});
		});
		const config = ClientConfig.create({
			endpoint: "http://127.0.0.1:9443",
			environment: Environment.Local,
			identity: { principalId: PRINCIPAL, projectId: PROJECT, tenantId: TENANT },
			insecureLoopbackForTesting: true,
			retry: { initialBackoffMs: 1, maxAttempts: 1, maxBackoffMs: 1 },
		});
		const client = MindcladeClient.withTransport(config, transport, runtime);
		const command = create(RequestJobCommandSchema, {
			configuration: artifact,
			context: { tenantId: "forged" },
			jobKind: "training",
			requestedJobId: "job-1",
		});
		await client.jobs.request(command, { idempotencyKey: "request-job" });
		await client.jobs.get("job-1");
		await client.jobs.list(create(ListJobsRequestSchema, { page: { pageSize: 25 } }));
		await client.jobs.cancel(
			create(CancelJobRequestSchema, { etag: "job-etag-1", name: "job-1", reason: "test" }),
			{ idempotencyKey: "cancel-job" },
		);
		await client.runs.getRun("run-1");
		await client.runs.listRuns(create(ListRunsRequestSchema, { parent: "job-1" }));
		await client.runs.getAttempt("attempt-1");
		await client.runs.listAttempts(create(ListAttemptsRequestSchema, { parent: "run-1" }));
		const acquisition = create(AcquireAttemptLeaseRequestSchema, {
			attemptId: "attempt-1",
			leaseDuration: { seconds: 120n },
			runName: "run-1",
		});
		const lease = await client.runs.acquire(acquisition, { idempotencyKey: "acquire" });
		assert.ok(lease.credential instanceof LeaseCredential);
		assert.equal(String(lease.credential), "LeaseCredential(<redacted>)");
		assert.doesNotMatch(String(lease.credential), new RegExp(TOKEN));
		const tampered = lease.attempt;
		tampered.attemptId = "attempts/tampered";
		assert.equal(lease.attempt.attemptId, attempt.attemptId);
		await client.runs.renew(
			create(RenewAttemptLeaseRequestSchema, {
				expectedResourceVersion: 1n,
				fence: lease.fence,
				leaseDuration: { seconds: 120n },
			}),
			lease.credential,
			{ idempotencyKey: "renew" },
		);
		await client.runs.heartbeat(
			create(HeartbeatAttemptRequestSchema, {
				expectedResourceVersion: 1n,
				fence: lease.fence,
				leaseDuration: { seconds: 120n },
			}),
			lease.credential,
			{ idempotencyKey: "heartbeat" },
		);
		await client.runs.cancelAttempt(
			create(CancelAttemptRequestSchema, {
				expectedResourceVersion: 1n,
				fence: lease.fence,
				reason: "worker shutdown",
			}),
			lease.credential,
			{ idempotencyKey: "cancel-attempt" },
		);
		const committedAttempt = create(AttemptSchema, lease.attempt);
		committedAttempt.state = AttemptState.SUCCEEDED;
		await client.runs.commitAttempt(
			create(CommitAttemptRequestSchema, {
				attempt: committedAttempt,
				expectedResourceVersion: 1n,
				fence: lease.fence,
				updateMask: { paths: ["state"] },
			}),
			lease.credential,
			{ idempotencyKey: "commit" },
		);

		assert.equal(command.context?.tenantId, "forged");
		assert.equal(acquisition.runName, "run-1");
		assert.deepEqual(calls, [
			"RequestJob",
			"GetJob",
			"ListJobs",
			"CancelJob",
			"GetRun",
			"ListRuns",
			"GetAttempt",
			"ListAttempts",
			"AcquireAttemptLease",
			"RenewAttemptLease",
			"HeartbeatAttempt",
			"CancelAttempt",
			"CommitAttempt",
		]);
		assert.deepEqual(leaseHeaders, [TOKEN, TOKEN, TOKEN, TOKEN]);
	});
});

File: sdks/go/mindclade/job_run_test.go
package mindclade

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"reflect"
	"strings"
	"sync"
	"testing"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/metadata"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/durationpb"
	"google.golang.org/protobuf/types/known/fieldmaskpb"
	"google.golang.org/protobuf/types/known/timestamppb"

	artifactv1 "github.com/mindclade/mindclade/protocols/generated/go/artifact/v1"
	commonv1 "github.com/mindclade/mindclade/protocols/generated/go/common/v1"
	internaljobv1 "github.com/mindclade/mindclade/protocols/generated/go/internalrpc/job/v1"
	jobv1 "github.com/mindclade/mindclade/protocols/generated/go/job/v1"
	operationv1 "github.com/mindclade/mindclade/protocols/generated/go/operation/v1"
)

// #nosec G101 -- deterministic non-secret capability fixture used only by an in-memory transport.
const jobRunLeaseToken = "opaque-worker-lease-capability-0123456789abcdef"

type jobRunFacadeClient struct {
	internaljobv1.JobServiceClient
	internaljobv1.RunServiceClient
	mu           sync.Mutex
	calls        []string
	leaseHeaders []string
}

func (client *jobRunFacadeClient) record(ctx context.Context, method string, commandContext *commonv1.CommandContext, message proto.Message, fenced bool) {
	client.mu.Lock()
	defer client.mu.Unlock()
	client.calls = append(client.calls, method)
	if commandContext != nil {
		if commandContext.GetTenantId() != "tenant-a" || commandContext.GetProjectId() != "project-a" || commandContext.GetPrincipalId() != "principal-a" || commandContext.GetCanonicalRequestDigest() == "" {
			panic("mutation command context was not bound to trusted client scope")
		}
		clone := proto.Clone(message)
		reflection := clone.ProtoReflect()
		reflection.Clear(reflection.Descriptor().Fields().ByName("context"))
		digest, err := deterministicDigest(clone)
		if err != nil || digest != commandContext.GetCanonicalRequestDigest() {
			panic("mutation command context did not carry the canonical request digest")
		}
	}
	requestValue, _ := ctx.Value(requestContextKey{}).(requestMetadata)
	if fenced {
		if requestValue.leaseToken != jobRunLeaseToken {
			panic("fenced mutation omitted its confidential transport credential")
		}
		client.leaseHeaders = append(client.leaseHeaders, requestValue.leaseToken)
	} else if requestValue.leaseToken != "" {
		panic("unfenced call leaked lease credential metadata")
	}
}

func (*jobRunFacadeClient) job() *jobv1.Job {
	return &jobv1.Job{JobId: "jobs/job-1", OperationId: "operations/op-1", TenantId: "tenant-a", ProjectId: "project-a", ResourceVersion: 1, State: jobv1.JobState_JOB_STATE_ACCEPTED}
}

func (*jobRunFacadeClient) operation() *operationv1.Operation {
	return &operationv1.Operation{OperationId: "operations/op-1", JobId: "jobs/job-1", TenantId: "tenant-a", ProjectId: "project-a", ResourceVersion: 1, State: operationv1.OperationState_OPERATION_STATE_PENDING}
}

func (*jobRunFacadeClient) run() *jobv1.Run {
	return &jobv1.Run{RunId: "runs/run-1", JobId: "jobs/job-1", TenantId: "tenant-a", ProjectId: "project-a", ResourceVersion: 1, LeaseEpoch: 1, State: jobv1.RunState_RUN_STATE_EXECUTING}
}

func (*jobRunFacadeClient) attempt(state jobv1.AttemptState) *jobv1.Attempt {
	return &jobv1.Attempt{AttemptId: "attempts/attempt-1", RunId: "runs/run-1", JobId: "jobs/job-1", TenantId: "tenant-a", ProjectId: "project-a", ResourceVersion: 1, LeaseEpoch: 1, State: state}
}

func (*jobRunFacadeClient) fence() *jobv1.LeaseFence {
	sum := sha256.Sum256([]byte(jobRunLeaseToken))
	return &jobv1.LeaseFence{AttemptId: "attempts/attempt-1", RunId: "runs/run-1", JobId: "jobs/job-1", TenantId: "tenant-a", ProjectId: "project-a", LeaseEpoch: 1, Deadline: timestamppb.New(time.Now().Add(5 * time.Minute)), LeaseTokenDigest: "sha256:" + hex.EncodeToString(sum[:])}
}

func (client *jobRunFacadeClient) RequestJob(ctx context.Context, request *internaljobv1.RequestJobRequest, _ ...grpc.CallOption) (*internaljobv1.RequestJobResponse, error) {
	client.record(ctx, "RequestJob", request.GetCommand().GetContext(), request.GetCommand(), false)
	return &internaljobv1.RequestJobResponse{Job: client.job(), Operation: client.operation()}, nil
}

func (client *jobRunFacadeClient) GetJob(ctx context.Context, request *internaljobv1.GetJobRequest, _ ...grpc.CallOption) (*internaljobv1.GetJobResponse, error) {
	client.record(ctx, "GetJob", nil, request, false)
	return &internaljobv1.GetJobResponse{Job: client.job()}, nil
}

func (client *jobRunFacadeClient) ListJobs(ctx context.Context, request *internaljobv1.ListJobsRequest, _ ...grpc.CallOption) (*internaljobv1.ListJobsResponse, error) {
	client.record(ctx, "ListJobs", nil, request, false)
	if request.GetParent() != "tenants/tenant-a/projects/project-a" {
		panic("ListJobs did not bind configured project scope")
	}
	return &internaljobv1.ListJobsResponse{Jobs: []*jobv1.Job{client.job()}, Page: &commonv1.PageResponse{NextPageToken: "jobs-next"}}, nil
}

func (client *jobRunFacadeClient) CancelJob(ctx context.Context, request *internaljobv1.CancelJobRequest, _ ...grpc.CallOption) (*internaljobv1.CancelJobResponse, error) {
	client.record(ctx, "CancelJob", request.GetContext(), request, false)
	return &internaljobv1.CancelJobResponse{Operation: client.operation()}, nil
}

func (client *jobRunFacadeClient) GetRun(ctx context.Context, request *internaljobv1.GetRunRequest, _ ...grpc.CallOption) (*internaljobv1.GetRunResponse, error) {
	client.record(ctx, "GetRun", nil, request, false)
	return &internaljobv1.GetRunResponse{Run: client.run()}, nil
}

func (client *jobRunFacadeClient) ListRuns(ctx context.Context, request *internaljobv1.ListRunsRequest, _ ...grpc.CallOption) (*internaljobv1.ListRunsResponse, error) {
	client.record(ctx, "ListRuns", nil, request, false)
	return &internaljobv1.ListRunsResponse{Runs: []*jobv1.Run{client.run()}, Page: &commonv1.PageResponse{NextPageToken: "runs-next"}}, nil
}

func (client *jobRunFacadeClient) GetAttempt(ctx context.Context, request *internaljobv1.GetAttemptRequest, _ ...grpc.CallOption) (*internaljobv1.GetAttemptResponse, error) {
	client.record(ctx, "GetAttempt", nil, request, false)
	return &internaljobv1.GetAttemptResponse{Attempt: client.attempt(jobv1.AttemptState_ATTEMPT_STATE_LEASED)}, nil
}

func (client *jobRunFacadeClient) ListAttempts(ctx context.Context, request *internaljobv1.ListAttemptsRequest, _ ...grpc.CallOption) (*internaljobv1.ListAttemptsResponse, error) {
	client.record(ctx, "ListAttempts", nil, request, false)
	return &internaljobv1.ListAttemptsResponse{Attempts: []*jobv1.Attempt{client.attempt(jobv1.AttemptState_ATTEMPT_STATE_LEASED)}, Page: &commonv1.PageResponse{NextPageToken: "attempts-next"}}, nil
}

func (client *jobRunFacadeClient) AcquireAttemptLease(ctx context.Context, request *internaljobv1.AcquireAttemptLeaseRequest, options ...grpc.CallOption) (*internaljobv1.AcquireAttemptLeaseResponse, error) {
	client.record(ctx, "AcquireAttemptLease", request.GetContext(), request, false)
	for _, option := range options {
		if header, ok := option.(grpc.HeaderCallOption); ok {
			*header.HeaderAddr = metadata.Pairs(leaseTokenHeaderSDK, jobRunLeaseToken)
		}
	}
	return &internaljobv1.AcquireAttemptLeaseResponse{Attempt: client.attempt(jobv1.AttemptState_ATTEMPT_STATE_LEASED), Fence: client.fence()}, nil
}

func (client *jobRunFacadeClient) RenewAttemptLease(ctx context.Context, request *internaljobv1.RenewAttemptLeaseRequest, _ ...grpc.CallOption) (*internaljobv1.RenewAttemptLeaseResponse, error) {
	client.record(ctx, "RenewAttemptLease", request.GetContext(), request, true)
	return &internaljobv1.RenewAttemptLeaseResponse{Attempt: client.attempt(jobv1.AttemptState_ATTEMPT_STATE_LEASED), Fence: client.fence()}, nil
}

func (client *jobRunFacadeClient) HeartbeatAttempt(ctx context.Context, request *internaljobv1.HeartbeatAttemptRequest, _ ...grpc.CallOption) (*internaljobv1.HeartbeatAttemptResponse, error) {
	client.record(ctx, "HeartbeatAttempt", request.GetContext(), request, true)
	return &internaljobv1.HeartbeatAttemptResponse{Attempt: client.attempt(jobv1.AttemptState_ATTEMPT_STATE_LEASED), Fence: client.fence(), ObservedAt: timestamppb.Now()}, nil
}

func (client *jobRunFacadeClient) CancelAttempt(ctx context.Context, request *internaljobv1.CancelAttemptRequest, _ ...grpc.CallOption) (*internaljobv1.CancelAttemptResponse, error) {
	client.record(ctx, "CancelAttempt", request.GetContext(), request, true)
	return &internaljobv1.CancelAttemptResponse{Attempt: client.attempt(jobv1.AttemptState_ATTEMPT_STATE_CANCELLED), Run: client.run()}, nil
}

func (client *jobRunFacadeClient) CommitAttempt(ctx context.Context, request *internaljobv1.CommitAttemptRequest, _ ...grpc.CallOption) (*internaljobv1.CommitAttemptResponse, error) {
	client.record(ctx, "CommitAttempt", request.GetContext(), request, true)
	return &internaljobv1.CommitAttemptResponse{Attempt: client.attempt(jobv1.AttemptState_ATTEMPT_STATE_SUCCEEDED), Run: client.run()}, nil
}

func TestJobRunFacadesCoverAllErgonomicRPCsAndHideLeaseCredentials(t *testing.T) {
	client, _, _ := testClient(t)
	transport := &jobRunFacadeClient{}
	client.Jobs.transport, client.Runs.transport = transport, transport
	ctx := context.Background()
	mutation := []RequestOption{WithIdempotencyKey("job-run-idempotency")}
	command := &jobv1.RequestJobCommand{Context: &commonv1.CommandContext{TenantId: "forged"}, JobKind: "training", Configuration: &artifactv1.ArtifactRef{Digest: "sha256:" + strings.Repeat("a", 64), MediaType: "application/json", SizeBytes: 42}, RequestedJobId: "job-1"}
	job, operation, err := client.Jobs.Request(ctx, command, mutation...)
	if err != nil || job.GetOperationId() != operation.GetOperationId() || command.GetContext().GetTenantId() != "forged" {
		t.Fatalf("RequestJob: job=%v operation=%v command=%v err=%v", job, operation, command, err)
	}
	if _, err = client.Jobs.Get(ctx, "job-1", ""); err != nil {
		t.Fatalf("GetJob: %v", err)
	}
	if _, err = client.Jobs.List(ctx, &internaljobv1.ListJobsRequest{Page: &commonv1.PageRequest{PageSize: 25}}); err != nil {
		t.Fatalf("ListJobs: %v", err)
	}
	if _, err = client.Jobs.Cancel(ctx, &internaljobv1.CancelJobRequest{Name: "job-1", Etag: "etag-1", Reason: "test"}, mutation...); err != nil {
		t.Fatalf("CancelJob: %v", err)
	}
	if _, err = client.Runs.GetRun(ctx, "run-1"); err != nil {
		t.Fatalf("GetRun: %v", err)
	}
	if _, err = client.Runs.ListRuns(ctx, &internaljobv1.ListRunsRequest{Parent: "job-1"}); err != nil {
		t.Fatalf("ListRuns: %v", err)
	}
	if _, err = client.Runs.GetAttempt(ctx, "attempt-1"); err != nil {
		t.Fatalf("GetAttempt: %v", err)
	}
	if _, err = client.Runs.ListAttempts(ctx, &internaljobv1.ListAttemptsRequest{Parent: "run-1"}); err != nil {
		t.Fatalf("ListAttempts: %v", err)
	}
	grant, err := client.Runs.AcquireAttemptLease(ctx, &internaljobv1.AcquireAttemptLeaseRequest{RunName: "run-1", AttemptId: "attempt-1", LeaseDuration: durationpb.New(2 * time.Minute)}, mutation...)
	if err != nil || grant.Credential().String() != "LeaseCredential(<redacted>)" {
		t.Fatalf("AcquireAttemptLease: grant=%v err=%v", grant, err)
	}
	detached := grant.Attempt()
	detached.AttemptId = "attempts/tampered"
	if grant.Attempt().GetAttemptId() != "attempts/attempt-1" {
		t.Fatal("LeaseGrant retained a caller-mutable attempt alias")
	}
	fence, credential := grant.Fence(), grant.Credential()
	if _, err = client.Runs.RenewAttemptLease(ctx, &internaljobv1.RenewAttemptLeaseRequest{Fence: cloneGenerated(fence), LeaseDuration: durationpb.New(2 * time.Minute), ExpectedResourceVersion: 1}, credential, mutation...); err != nil {
		t.Fatalf("RenewAttemptLease: %v", err)
	}
	if _, err = client.Runs.HeartbeatAttempt(ctx, &internaljobv1.HeartbeatAttemptRequest{Fence: cloneGenerated(fence), LeaseDuration: durationpb.New(2 * time.Minute), ExpectedResourceVersion: 1}, credential, mutation...); err != nil {
		t.Fatalf("HeartbeatAttempt: %v", err)
	}
	if _, err = client.Runs.CancelAttempt(ctx, &internaljobv1.CancelAttemptRequest{Fence: cloneGenerated(fence), ExpectedResourceVersion: 1, Reason: "worker shutdown"}, credential, mutation...); err != nil {
		t.Fatalf("CancelAttempt: %v", err)
	}
	if _, err = client.Runs.CommitAttempt(ctx, &internaljobv1.CommitAttemptRequest{Attempt: transport.attempt(jobv1.AttemptState_ATTEMPT_STATE_SUCCEEDED), Fence: cloneGenerated(fence), UpdateMask: &fieldmaskpb.FieldMask{Paths: []string{"state"}}, ExpectedResourceVersion: 1}, credential, mutation...); err != nil {
		t.Fatalf("CommitAttempt: %v", err)
	}
	want := []string{"RequestJob", "GetJob", "ListJobs", "CancelJob", "GetRun", "ListRuns", "GetAttempt", "ListAttempts", "AcquireAttemptLease", "RenewAttemptLease", "HeartbeatAttempt", "CancelAttempt", "CommitAttempt"}
	transport.mu.Lock()
	defer transport.mu.Unlock()
	if !reflect.DeepEqual(transport.calls, want) {
		t.Fatalf("calls=%v want=%v", transport.calls, want)
	}
	if !reflect.DeepEqual(transport.leaseHeaders, []string{jobRunLeaseToken, jobRunLeaseToken, jobRunLeaseToken, jobRunLeaseToken}) {
		t.Fatalf("lease headers were not isolated to fenced calls: %v", transport.leaseHeaders)
	}
}
